### PR TITLE
[2.8] Support dynamic instance type discovery on GCE

### PR DESCRIPTION
**Pull request by achilleasa**
This PR removes the hard-coded instance type list from the GCE provider and populates the instance type list on demand using the GCE API.

A potential caveat of this change (as you can see in the QA step output below) is that juju will request a `n1-highcpu-4` machine for the controller (to match the implicit memory constraints for controllers) **and** `e2-micro` machines for deployed apps (see https://cloud.google.com/compute/docs/machine-types). Before this change, we would pick a `n1-standard-1` instance for deployed charms which has more available memory (although the micro ones can potentially scale up their memory using the [virtio baloon](https://cloud.google.com/compute/docs/machine-types#virtio_memory_balloon_device) device.

## QA steps

```sh
$ juju bootstrap google gce --no-gui
$ juju deploy cs:wordpress
$ juju deploy cs:percona-cluster --constraints 'instance-type=e2-standard-2'

# Wait for the machines to come up; then:
$ gcloud compute instances list
NAME                                                 ZONE               MACHINE_TYPE   PREEMPTIBLE  INTERNAL_IP  EXTERNAL_IP     STATUS
juju-f0b4a3-0                                        us-east1-b         n1-highcpu-4                10.240.0.8   35.185.37.187   RUNNING
juju-b8dd2a-0                                        us-east1-b         e2-micro                    10.240.0.46  35.231.173.30   RUNNING
juju-b8dd2a-1                                        us-east1-c         e2-standard-2               10.240.0.71  34.73.96.116    RUNNING
```

## Bug reference
https://bugs.launchpad.net/juju/+bug/1918233

### CODE REVIEW

**Comment by hmlanigan** [comment]
@achilleasa, could you please PM me how you got the gcloud login working with our creds?

**Review by hmlanigan — COMMENTED** [review]
LGTM, 2 nits.  The juju piece of QA went well, however I haven't been able to login with gcloud yet to verify the instance types used.

**Comment by wallyworld** [comment]
LIke was done for AWS when dynamic instance queries were added, we'll need to add caching on the environ instance. Otherwise I suspect during a normal bootstrap process we'll be making several queries to seed the constraints vocab, do the prechecks, start the instance etc etc.
Maybe we could start by logging each time a call is made during bootstrap and deploy operations and verify that my suspicions are correct and then add the caching to address the issue.

File: provider/gce/instance_information.go [hunk 1]
```diff
@@ -16,8 +16,12 @@ import (
 	"github.com/juju/juju/provider/gce/google"
 )
 
-var _ environs.InstanceTypesFetcher = (*environ)(nil)
-var virtType = "kvm"
+var (
+	_ environs.InstanceTypesFetcher = (*environ)(nil)
+
+	virtType   = "kvm"
+	machArches = []string{arch.AMD64}
```
[the remark this answers is not in the document]
Comment (achilleasa) [reply provider/gce/instance_information.go:23]: No idea; I left it as-is thinking that there must be there for some (albeit undocumented :-() reason. I will play around with the gcloud cli to see what instance types we get back, update `machArches` accordingly and run a few more tests.

[the remark this answers is not in the document]
Comment (achilleasa) [reply provider/gce/instance_information.go:23]: Look's like there's only amd64 support for now: https://cloud.google.com/compute/docs/cpu-platforms. I will add a comment as a reminder.

**Comment by achilleasa** [comment]
@wallyworld I have added a cache with a 10min expiration policy. If that looks OK to you, I can also apply this (via a separate PR) to the [ec2](https://github.com/juju/juju/blob/develop/provider/ec2/instancetypes.go#L179) one.

**Review by hmlanigan — APPROVED** [review]
LGTM.

I couldn't get credentials for gcloud to work.  However deploy went well.  The 'e2-standard-2' instance is not in the hard coded list of instances, and still deployed well.

File: provider/gce/instance_information.go [hunk 2]
```diff
@@ -16,18 +18,48 @@ import (
...
+// getAllInstanceTypes fetches and memoizes the list of available GCE instances
+// for the AZs associated with the current region.
+func (env *environ) getAllInstanceTypes(ctx context.ProviderCallContext, clock clock.Clock) ([]instances.InstanceType, error) {
+	env.instTypeListLock.Lock()
+	defer env.instTypeListLock.Unlock()
+
+	if len(env.cachedInstanceTypes) != 0 && clock.Now().Before(env.instCacheExpireAt) {
+		return env.cachedInstanceTypes, nil
+	}
+
```
[the remark this answers is not in the document]
Comment (achilleasa) [reply provider/gce/instance_information.go:55]: I am thinking of relocating this logic into the provider common package so we can use it consistently across all providers. I will make a note to add a trace in that PR.

**Comment by achilleasa** [comment]
$$merge$$

**Pull request merged**